### PR TITLE
Change query field expansion

### LINKED ISSUE #31798: multi_match query with wildcard against boolean doesn't expand
**Elasticsearch version** : 6.3.0
**JVM version** : 1.8.0_144
**OS version** : OSX

**To Reproduce:**

1) Index various data types:

```
PUT test3/doc/1
{
  "obj": {
    "flag": true,
    "text": "asdf",
    "numeric": 42
  }
}
```

2) Search the boolean value with wildcard in field name:

```
GET /test3/_search
{
  "query": {
    "multi_match": {
      "query":    true,
      "fields":   [ "*.flag" ]
    }
  }
}
```

The above fails, even though the same search without expansion succeeds:

```
GET /test3/_search
{
  "query": {
    "multi_match": {
      "query":    true,
      "fields":   [ "obj.flag" ]
    }
  }
}
```

Text or Numeric searches with expansion succeed:

```
GET /test3/_search
{
  "query": {
    "multi_match": {
      "query":    42,
      "fields":   [ "*.numeric" ]
    }
  }
}
```

### PR DESCRIPTION
**Pull request by jimczi**
This commit changes the query field expansion for query parsers
to not rely on an hardcoded list of field types. Instead we rely on
the type of exception that is thrown by MappedFieldType#termQuery to
include/exclude an expanded field.

Supersedes #31655

Closes #31798

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java [hunk 1]
```diff
@@ -177,18 +132,20 @@ public static Mapper getFieldMapper(MapperService mapperService, String field) {
...
-            if (acceptMetadataField == false && mapper instanceof MetadataFieldMapper) {
-                continue;
+            if (acceptAllTypes == false) {
+                try {
+                    fieldType.termQuery("", context);
+                } catch (QueryShardException |UnsupportedOperationException e) {
+                    // field type is never searchable with term queries (eg. geo point): ignore
+                    continue;
+                } catch (RuntimeException e) {
+                    // other exceptions are parsing errors or not indexed fields: keep
```
Comment (colings86) [inline server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java]: Should we not be throwing the exception here and reporting the problem back to the user?

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java]: No, this case is for field mappers that need a special input (numbers, boolean, ...) so this is consistent with the current behavior. The goal here is to select fields that can handle a `term` query so we silently ignore those that can't. Previously this selection was based on an hardcoded list of field types.

Comment (jpountz) [reply server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java]: can we narrow down the exception type here, eg. IAE?

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java]: True thanks, it should be IAE only. I pushed the change.

**Review by jpountz — APPROVED** [review]

**Pull request merged**